bees/execbee/execbee.go

```diff
@@ -0,0 +1,120 @@
+/*
+ *    Copyright (C) 2015 Dominik Schmidt
+ *
+ *    This program is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU Affero General Public License as published
+ *    by the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    This program is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ *    Authors:
+ *      Dominik Schmidt <domme@tomahawk-player.org>
+ */
+
+// beehive's IRC module.
+package execbee
+
+import (
+	"bufio"
+	"fmt"
+	"github.com/muesli/beehive/bees"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+)
+
+type ExecBee struct {
+	bees.Bee
+
+	eventChan chan bees.Event
+}
+
+// Interface impl
+func (mod *ExecBee) Action(action bees.Action) []bees.Placeholder {
+	outs := []bees.Placeholder{}
+
+	switch action.Name {
+	case "localCommand":
+		for _, opt := range action.Options {
+			if opt.Name == "command" {
+				log.Println("Execute locally: ", opt.Name, opt.Value.(string))
+
+				go func() {
+					c := strings.Split(opt.Value.(string), " ")
+					cmd := exec.Command(c[0], c[1:]...)
+
+					// read and print stdout
+					outReader, err := cmd.StdoutPipe()
+					if err != nil {
+						fmt.Fprintln(os.Stderr, "Error creating StdoutPipe for Cmd", err)
+						return
+					}
+					outBuffer := []string{}
+					outScanner := bufio.NewScanner(outReader)
+					go func() {
+						for outScanner.Scan() {
+							foo := outScanner.Text()
+							log.Println("execbee: std: | ", foo)
+							outBuffer = append(outBuffer, foo)
+						}
+					}()
+
+					// read and print stderr
+					errReader, err := cmd.StderrPipe()
+					if err != nil {
+						fmt.Fprintln(os.Stderr, "Error creating StderrPipe for Cmd", err)
+						return
+					}
+					errBuffer := []string{}
+					errScanner := bufio.NewScanner(errReader)
+					go func() {
+						for errScanner.Scan() {
+							foo := errScanner.Text()
+							log.Println("execbee: err: | ", foo)
+							errBuffer = append(errBuffer, foo)
+						}
+					}()
+
+					cmd.Start()
```

Comment:
Start may fail - e.g. if a command doesn't exist. Should check for error returned for Start() and Wait() below, too.